### ISSUE TITLE
add support for setting JSX for title on Dialog props

### ISSUE THREAD

**Issue by shiftkey** [desktop/desktop#6032]
We want to avoid contributors using `Dialog` and nesting `DialogHeader` within if they only want the default behaviour. See https://github.com/desktop/desktop/pull/6027#discussion_r229337387 for more discussion.

**Comment by shiftkey** [comment]
I was going to propose using `react-to-string` here which seems to fit the bill, based on the example:

```js
import React, { Component } from 'react';
import reactToString from 'react-to-string';
 
class App extends Component {
  render() {
    return <div>Hello world!</div>
  }
}
 
reactToString(<App />);
// Returns "Hello world!"
 
reactToString(<p>Are you interested in this library? <a href="signup">Click here!</a></p>);
// Returns "Are you interested in this library? Click here!"
```

But the library does not have a license, so I've opened https://github.com/kevinzwhuang/react-to-string/issues/3 to ask them to define one before I look at pulling this into our codebase.

**Comment by shiftkey** [comment]
> It looks like that doesn’t actually _render_ the element so the code might not work as described.

To be clear,  I'm only looking for a string-like representation of the element for situations where we need something distinct to work with (the original issue was using the title to generate a unique id). I'm not looking to actually render the elements.

**Comment by shiftkey** [comment]
This was the snippet I used in #6027 as a starting point:

```diff
diff --git a/app/src/ui/dialog/dialog.tsx b/app/src/ui/dialog/dialog.tsx
index 63162f312..9ff9cdee7 100644
--- a/app/src/ui/dialog/dialog.tsx
+++ b/app/src/ui/dialog/dialog.tsx
@@ -31,7 +31,7 @@ interface IDialogProps {
    * By omitting this consumers may use their own custom DialogHeader
    * for when the default component doesn't cut it.
    */
-  readonly title?: string
+  readonly title?: string | JSX.Element
 
   /**
    * Whether or not the dialog should be dismissable. A dismissable dialog
@@ -178,9 +178,11 @@ export class Dialog extends React.Component<IDialogProps, IDialogState> {
       this.setState({ titleId: undefined })
     }
 
-    if (this.props.title) {
+    if (this.props.title != null) {
+      // TODO: how can we get a string-like representation of a JSX element?
+      const id = typeof this.props.title === 'string' ? this.props.title : '???'
       this.setState({
-        titleId: createUniqueId(`Dialog_${this.props.id}_${this.props.title}`),
+        titleId: createUniqueId(`Dialog_${this.props.id}_${id}`),
       })
     }
   }
diff --git a/app/src/ui/merge-branch/merge.tsx b/app/src/ui/merge-branch/merge.tsx
index bd4239bc7..0d0ed27a3 100644
--- a/app/src/ui/merge-branch/merge.tsx
+++ b/app/src/ui/merge-branch/merge.tsx
@@ -19,7 +19,6 @@ import { MergeResultKind } from '../../models/merge'
 import { MergeStatusHeader } from '../history/merge-status-header'
 import { promiseWithMinimumTimeout } from '../../lib/promise'
 import { truncateWithEllipsis } from '../../lib/truncate-with-ellipsis'
-import { DialogHeader } from '../dialog/header'
 
 interface IMergeProps {
   readonly dispatcher: Dispatcher
@@ -293,18 +292,15 @@ export class Merge extends React.Component<IMergeProps, IMergeState> {
     return (
       <Dialog
         id="merge"
+        title={
+          <div className="merge-dialog-header">
+            Merge into <b>{currentBranchName}</b>
+          </div>
+        }
         onDismissed={this.props.onDismissed}
+        dismissable={true}
         onSubmit={this.merge}
       >
-        <DialogHeader
-          title={
-            <div className={'merge-dialog-header'}>
-              Merge into <b>{currentBranchName}</b>
-            </div>
-          }
-          dismissable={true}
-          onDismissed={this.props.onDismissed}
-        />
         <DialogContent>
           <BranchList
             allBranches={this.props.allBranches}
```

**Comment by niik** [comment]
Hey @shiftkey, sorry, I've completed missed this.

> ```ts
> // TODO: how can we get a string-like representation of a JSX element?
>  const id = typeof this.props.title === 'string' ? this.props.title : '???'```

You can ignore this entirely. CreateUniqueId will ensure it's a unique id regardless of what you pass it so just leave that as-is (ie. pass the raw JSX.Element to it, that's fine). You can use `createUniqueId` without giving it anything if you like and it'll take care of it, we only give it something that's slightly human readable for sanity's sake.

**Comment by shiftkey** [comment]
@niik cool, that should simplify the other things in here

**Comment by shiftkey** [comment]
The commit to add this support is https://github.com/desktop/desktop/pull/6961/commits/6cf4a2aab1a4a648c7becdd1c4b6777bd1534bb2 and this should be resolved in #6961.

**Comment by shiftkey** [comment]
Resolved in #6961, and aside from the usage reported in #7043 there are two others:

 - rebase conflicts dialog

https://github.com/desktop/desktop/blob/b5529325ae629a52b02cb3e07699f1ccbcfb2cfa/app/src/ui/rebase/rebase-conflicts-dialog.tsx#L167-L171

 - merge branch selector

https://github.com/desktop/desktop/blob/b5529325ae629a52b02cb3e07699f1ccbcfb2cfa/app/src/ui/merge-branch/merge.tsx#L262-L271

 - release notes dialog

https://github.com/desktop/desktop/blob/b5529325ae629a52b02cb3e07699f1ccbcfb2cfa/app/src/ui/release-notes/release-notes-dialog.tsx#L159-L175

I'll move these out to `help wanted` tasks that can be picked up by others before closing this out.

**Comment by shiftkey** [comment]
Closing this out as the remaining work to cleanup this anti-pattern has been written up
